src/ui/public/state_management/state_storage/_tests__/hashed_item_store.js

```diff
@@ -0,0 +1,285 @@
+
+      expect(hashedItemStore._itemIndex).to.eql({
+        [persistedItemIndex.a.hash]: persistedItemIndex.a,
+        [persistedItemIndex.b.hash]: persistedItemIndex.b,
+      });
+    });
+  });
+
+  describe('#setItem', () => {
+    describe('if the item exists in sessionStorage', () => {
+      let sessionStorage;
+      let hashedItemStore;
+      let result;
+      const hash = 'a';
+      const item = JSON.stringify({});
+
+      beforeEach(() => {
+        sessionStorage = new StubBrowserStorage();
+        hashedItemStore = new HashedItemStore(sessionStorage);
+        result = hashedItemStore.setItem(hash, item);
+      });
+
+      it('updates the indexed item as recently touched', done => {
+        const prevTouched = hashedItemStore._itemIndex.a.touched;
+        setTimeout(() => {
+          try {
+            hashedItemStore.setItem(hash, item);
+            const touched = hashedItemStore._itemIndex.a.touched;
+            expect(touched).to.be.above(prevTouched);
+          } catch (e) {
+            return done(e);
+          }
+          done();
+        }, 1);
+      });
+
+      it('persists the item in sessionStorage', () => {
+        expect(sessionStorage.getItem(hash)).to.equal(item);
+      });
+
+      it('returns true', () => {
+        expect(result).to.equal(true);
+      });
+    });
+
+    describe(`if the item doesn't exist in sessionStorage`, () => {
+      describe(`if there's storage space`, () => {
+        let sessionStorage;
+        let hashedItemStore;
+        const hash = 'a';
+        const item = JSON.stringify({});
+
+        beforeEach(() => {
+          sessionStorage = new StubBrowserStorage();
+          hashedItemStore = new HashedItemStore(sessionStorage);
+        });
+
+        it('updates the indexed item as recently touched', done => {
+          const time = Date.now();
+          setTimeout(() => {
+            try {
+              hashedItemStore.setItem(hash, item);
+              const touched = hashedItemStore._itemIndex.a.touched;
+              expect(touched).to.be.above(time);
+            } catch (e) {
+              return done(e);
+            }
+            done();
+          }, 1);
+        });
+
+        it('persists the item in sessionStorage', () => {
+          hashedItemStore.setItem(hash, item);
+          expect(sessionStorage.getItem(hash)).to.equal(item);
+        });
+
+        it('returns true', () => {
+          const result = hashedItemStore.setItem(hash, item);
+          expect(result).to.equal(true);
+        });
+      });
+
+      describe(`if there isn't storage space`, () => {
+        let sessionStorage;
+        let hashedItemStore;
+        let storageSizeLimit;
+        const hash = 'a';
+        const item = JSON.stringify({});
+
+        function calculateItemSize(itemCount) {
+          const itemSize = hash.length + item.length;
+          return itemSize * itemCount;
+        }
+
+        function calculateIndexSize(itemCount) {
```

Comment:
I don't think it would be a bad idea to add a method to the HashedItemStore that would gives you the serialized size of a set of items, including the index, that way at least the implementation and the implementation specific test helpers would be in the same place.